_bleio/service.py

```diff
@@ -0,0 +1,116 @@
+# of this software and associated documentation files (the "Software"), to deal
+# in the Software without restriction, including without limitation the rights
+# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+# copies of the Software, and to permit persons to whom the Software is
+# furnished to do so, subject to the following conditions:
+#
+# The above copyright notice and this permission notice shall be included in
+# all copies or substantial portions of the Software.
+#
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+# THE SOFTWARE.
+"""
+`_bleio.service`
+=======================================================================
+
+_bleio implementation for Adafruit_Blinka_bleio
+
+* Author(s): Dan Halbert for Adafruit Industries
+"""
+from __future__ import annotations
+from typing import Tuple, Union
+
+from bleak.backends.service import BleakGATTService
+
+from _bleio.characteristic import Characteristic
+import _bleio.connection
+from _bleio.uuid_ import UUID
+
+
+class Service:
+    """Stores information about a BLE service and its characteristics."""
+
+    def __init__(
+        self, uuid: UUID, *, secondary: bool = False, remote: bool = False,
+    ):
+        """Create a new Service identified by the specified UUID. It can be accessed by all
+        connections. This is known as a Service server. Client Service objects are created via
+        `Connection.discover_remote_services`.
+
+        To mark the Service as secondary, pass `True` as :py:data:`secondary`.
+
+        :param UUID uuid: The uuid of the service
+        :param bool secondary: If the service is a secondary one
+
+        :return: the new Service
+        """
+        self._uuid = uuid
+        self._secondary = secondary
+        self._remote = remote
+        self._connection = None
+        self._characteristics = ()
+        self._bleak_gatt_service = None
+
+    # pylint: disable=protected-access
+    @classmethod
+    def from_bleak(
+        cls,
+        connection: _bleio.connection.Connection,
+        bleak_gatt_service: BleakGATTService,
+    ) -> Service:
+        service = cls(UUID(bleak_gatt_service.uuid), remote=True)
+        service._connection = connection
+        service._characteristics = tuple(
+            Characteristic.from_bleak(service, bleak_characteristic)
+            for bleak_characteristic in bleak_gatt_service.characteristics
+        )
+        service._bleak_gatt_service = bleak_gatt_service
+        return service
+
+    @property
+    def bleak_service(self):
```

Comment:
Again, why is this public? (Same question for `from_bleak`.)

Reply:
It needed to be called from other classes, it's not internal to `Service only`. Basically it is like a C++ "friend" method. If I underscore it, I'll have to pylint-disable a lot of places. I guess that's the other choice, so I'll do taht.